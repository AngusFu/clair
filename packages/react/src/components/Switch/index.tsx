import classNames from "classnames";
import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import { AutoIncreasingCounter } from "@clair/helpers";
import { isColorValid } from "../Tag/index";

const autoIncrCounter = new AutoIncreasingCounter();

const sizes = ["normal", "small"] as const;
export type Size = typeof sizes[number];
const PSize = PropTypes.oneOf([...sizes]);

type Text = React.ReactNode;
const PText = PropTypes.node;

type EventHandler = (event: CFormEvent) => void;

export interface SwitchProps {
  defaultChecked?: boolean;
  checked?: boolean;
  disabled?: boolean;
  checkedChildren?: Text;
  unCheckedChildren?: Text;
  size?: Size;
  checkedColor?: string;
  unCheckedColor?: string;
  onChange?: EventHandler;
}

export const Switch: React.FC<SwitchProps> = props => {
  const {
    defaultChecked = false,
    disabled = false,
    checkedChildren,
    unCheckedChildren,
    size = "normal",
    checkedColor,
    unCheckedColor,
    onChange = null
  } = props;

  const className: string = classNames(
    "c-switch",
    size !== "normal" ? `c-switch--${size}` : null
  );

  const isControlled = "checked" in props;
  const [defChecked, setChecked] = useState(defaultChecked);
  const checked = isControlled ? props.checked : defChecked;

  const handleChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
    const isChecked = !checked;
    if (!isControlled) {
      setChecked(isChecked);
    }
    const event: CFormEvent = {
      target: {
        value: isChecked
      },
      nativeEvent: e.nativeEvent
    };
    typeof onChange === "function" && onChange(event);
  };

  const getStyles = (color: string | undefined) => {
    return color && isColorValid(color) ? { background: color } : {};
  };

  const styles: object = checked
    ? getStyles(checkedColor)
    : getStyles(unCheckedColor);

  const hasChildren = Boolean(checkedChildren && unCheckedChildren);

  const labelIdRef = useRef(`c-switch__label-${autoIncrCounter.next()}`);
  const [labelId, updateLabelId] = useState<string | undefined>(
    labelIdRef.current
  );

  useEffect(() => {
    updateLabelId(() => {
      return hasChildren ? labelIdRef.current : undefined;
    });
  }, [hasChildren]);

  const label = checked ? "开" : "关";
  const ariaLabel = hasChildren ? undefined : label;

  return (
    <label
      className={className}
      // eslint-disable-next-line
      role="switch"
      aria-checked={checked}
      aria-labelledby={labelId}
      aria-label={ariaLabel}
    >
      <input
        type="checkbox"
        className="c-switch__input"
        checked={checked}
        disabled={disabled}
        onChange={handleChange}
      />
      <span className="c-switch__box" style={styles} />
      <span className="c-switch__label" id={labelId}>
        {checked ? checkedChildren : unCheckedChildren}
      </span>
    </label>
  );
};

Switch.propTypes = {
  defaultChecked: PropTypes.bool,
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  checkedChildren: PText,
  unCheckedChildren: PText,
  size: PSize,
  checkedColor: PropTypes.string,
  unCheckedColor: PropTypes.string,
  onChange: PropTypes.func
};

Switch.displayName = "Switch";
