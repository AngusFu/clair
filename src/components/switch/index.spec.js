import { mount, shallowMount } from '@vue/test-utils'
import Switch from '@/components/switch'

const VModelSwitch = {
  template: `
<c-switch
  checked-color="#C7543A"
  unchecked-color="#E9CD4C"
  v-model="checked"
/>`,
  components: {
    'c-switch': Switch
  },
  data () {
    return {
      checked: true
    }
  }
}

it('should trigger v-model changes', () => {
  const wrapper = mount(VModelSwitch)
  expect(wrapper.vm.checked).toEqual(true)

  wrapper.find('.c-switch__layoutbox').trigger('click')
  expect(wrapper.vm.checked).toEqual(false)

  expect(wrapper).toMatchSnapshot()
})

it('should not trigger changes if disabled', () => {
  const wrapper = shallowMount(Switch, {
    propsData: { disabled: true }
  })

  // before click: false
  expect(wrapper.vm.currentValue).toEqual(false)
  // trigger click
  wrapper.find('.c-switch__layoutbox').trigger('click')
  // after click: false
  expect(wrapper.vm.currentValue).toEqual(false)

  expect(wrapper).toMatchSnapshot()
})

it('should use given value instead of boolean values', () => {
  const wrapper = shallowMount(Switch, {
    propsData: {
      value: 'checked',
      checkedValue: 'checked',
      uncheckedValue: 'unchecked'
    }
  })

  wrapper.find('.c-switch__layoutbox').trigger('click')
  expect(wrapper.vm.currentValue).toEqual('unchecked')
  wrapper.find('.c-switch__layoutbox').trigger('click')
  expect(wrapper.vm.currentValue).toEqual('checked')

  expect(wrapper).toMatchSnapshot()
})

it('should transform undefined values to boolean', () => {
  const wrapper = shallowMount(Switch, {
    propsData: {
      value: 'checked',
      checkedValue: 'checked',
      uncheckedValue: undefined
    }
  })

  wrapper.find('.c-switch__layoutbox').trigger('click')
  expect(wrapper.vm.currentValue).toEqual(false)
  wrapper.find('.c-switch__layoutbox').trigger('click')
  expect(wrapper.vm.currentValue).toEqual('checked')

  expect(wrapper).toMatchSnapshot()
})
