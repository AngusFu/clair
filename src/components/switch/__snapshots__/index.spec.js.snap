// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not trigger changes if disabled 1`] = `
<div class="c-switch"><input type="checkbox" disabled="disabled" name="c-switch" class="c-switch__checkbox" value="false">
  <div class="c-switch__layoutbox"></div>
</div>
`;

exports[`should transform undefined values to boolean 1`] = `
<div class="c-switch"><input type="checkbox" name="c-switch" class="c-switch__checkbox" value="checked">
  <div class="c-switch__layoutbox"></div>
</div>
`;

exports[`should trigger v-model changes 1`] = `
<div class="c-switch"><input type="checkbox" name="c-switch" class="c-switch__checkbox" value="false">
  <div class="c-switch__layoutbox" style="background-color: rgb(233, 205, 76); border-color: #e9cd4c;"></div>
</div>
`;

exports[`should use given value instead of boolean values 1`] = `
<div class="c-switch"><input type="checkbox" name="c-switch" class="c-switch__checkbox" value="checked">
  <div class="c-switch__layoutbox"></div>
</div>
`;
